src/main/scala/org/allenai/pipeline/S3Artifact.scala

```diff
@@ -1,23 +1,26 @@
 package org.allenai.pipeline
 
 import org.allenai.common.Logging
-import org.allenai.pipeline
 
 import com.amazonaws.AmazonServiceException
 import com.amazonaws.auth.BasicAWSCredentials
 import com.amazonaws.services.s3.AmazonS3Client
-import com.amazonaws.services.s3.model.{ CannedAccessControlList, PutObjectRequest, ObjectMetadata }
+import com.amazonaws.services.s3.model.{ CannedAccessControlList, ObjectMetadata, PutObjectRequest }
 
-import java.io.{ InputStream, File, FileOutputStream }
+import java.io.{ File, FileOutputStream, InputStream }
 import java.net.URI
 
-case class S3Config(service: AmazonS3Client, bucket: String)
+case class S3Config(initService: () => AmazonS3Client, bucket: String) {
+  @transient lazy val service = initService()
+}
 
 object S3Config {
-  def apply(accessKey: String, secretAccessKey: String, bucket: String): S3Config =
-    S3Config(new AmazonS3Client(new BasicAWSCredentials(accessKey, secretAccessKey)), bucket)
-
-  def apply(bucket: String): S3Config = S3Config(new AmazonS3Client(), bucket)
+  def apply(accessKey: String, secretAccessKey: String, bucket: String): S3Config = {
+    S3Config(() => new AmazonS3Client(new BasicAWSCredentials(accessKey, secretAccessKey)), bucket)
+  }
+  def apply(bucket: String): S3Config = {
+    S3Config(() => new AmazonS3Client(), bucket)
```

Comment:
so what's the final story on S3 creds?

On Mon, Mar 30, 2015 at 11:53 AM, cristipp <notifications@github.com> wrote:

> In src/main/scala/org/allenai/pipeline/S3Artifact.scala
> <https://github.com/allenai/pipeline/pull/21#discussion_r27420486>:
>
> >
> >  object S3Config {
> > -  def apply(accessKey: String, secretAccessKey: String, bucket: String): S3Config =
> > -    S3Config(new AmazonS3Client(new BasicAWSCredentials(accessKey, secretAccessKey)), bucket)
> > -
> > -  def apply(bucket: String): S3Config = S3Config(new AmazonS3Client(), bucket)
> > +  def apply(accessKey: String, secretAccessKey: String, bucket: String): S3Config = {
> > +    S3Config(() => new AmazonS3Client(new BasicAWSCredentials(accessKey, secretAccessKey)), bucket)
> > +  }
> > +  def apply(bucket: String): S3Config = {
> > +    S3Config(() => new AmazonS3Client(), bucket)
>
> Yep, this is where it's at. Just make the client already :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/allenai/pipeline/pull/21/files#r27420486>.
>



-- 

  -jake